src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -584,5 +585,7 @@ private void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
                 ThrowHelper.ThrowJsonException_SerializationConverterWrite(ConverterBase);
             }
         }
+
+        public ConcurrentDictionary<Type, JsonPropertyInfo> RuntimePropertyCache;
```

Comment:
Should this be stored on the `JsonSerializerOptions` instead with the rest of the caches?

Also, can the following existing cache be used for the runtime properties as well?

https://github.com/dotnet/corefx/blob/437d9bc76489ff18bd9ce04f287e07591f6bc38c/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L23

Reply:
On `_objectJsonProperties` I believe @layomia is working to get rid of that cache.

Reply:
> Should this be stored on the JsonSerializerOptions instead with the rest of the caches?

By storing the cache on the actual `JsonPropertyInfo` like it is currently means we are tied to a particular class + particular property that may have specific attribute (e.g. `[JsonIgnore]` and `[JsonConverter]` which we need to preserve.

If we store the cache on a `JsonSerializerOptions` instance then we would would need to key on class type + property name + property type.

If we store the cache on a `JsonClassInfo` instance then we would need to key on property name + property type.

The cache in the current location of a `JsonPropertyInfo` instance is the most specific, but that also means we are instantiating a new dictionary for every polymorphic property, which probably isn't ideal, especially in the future when we likely support polymorphic properties in cases other than System.Object.

So I think storing on `JsonClassInfo` is the best location like we have for the current property cache already there: `public volatile Dictionary<string, JsonPropertyInfo> PropertyCache`. However, we should be able to just use the existing cache by changing the "string" key to be property name + property type (just for polymorphic properties) then we can just re-use the existing cache.

Reply:
@steveharter 

> we should be able to just use the existing cache

I don't think that would be a good idea, since it would require allocating a string for every access to serve as the key. So you're likely paying lots of string allocations to avoid a single dictionary allocation.

Just moving the cache to `JsonClassInfo` (while using a (name, type) tuple as the key) sounds reasonable to me.

Reply:
@svick yes allocating a string for the key is not ideal, but small overhead compared to what we have today. A dictionary is heavy-weight so having one instance for a Type is much better than per property.

However creating a new cache on `JsonClassInfo` with the propertyName+propertyType tuple as you suggest would be OK, but JsonPropertyInfo+propertyType might be a bit more performant and works because the same JsonPropertyInfo coming in is (should) be the same instance each time.